lib/limits.c

```diff
@@ -62,8 +62,12 @@ static int setrlimit_value (unsigned int resource,
 		 * Also, we are limited to base 10 here (hex numbers will not
 		 * work with the limit string parser as is anyway)
 		 */
+		errno = 0;
 		l = strtol(value, &endptr, 10);
 
+		if (l == 0 && errno != 0)
```

Comment:
0 is a legitimate return value for `strtol()` on both success and failure (check man NOTES). I'd remove that part and leave the errno check (this is explained in the EXAMPLES section).

Reply:
I read the page, but I'm not sure it's entirely correct.  BTW, after reading again the page, I'm thinking this check is also superfluous.

strtol(3) can have the following cases (assuming a valid vase):

-  No conversion was performed, that is, no digits were found.  Returns 0, *endptr == str, and may set errno to EINVAL.

-  Overflow or underflow.  Returns LONG_MAX or LONG_MIN, _and_ sets errno to ERANGE.

-  Valid conversion.  Returns any value, possibly 0, LONG_MAX, or LONG_MIN.  Doesn't set errno to ERANGE nor EINVAL.  **But**, I don't see any guarantees that errno is not set; and all libc calls are allowed to set errno if they don't fail, which makes this very weird.

Since overflow or underflow give reasonable values that will later fail our __builtin_mul_overflow() check, we don't need to test for those.

And since we already test for *endptr == str, we're already testing when no conversion was performed.

So we were already covered by the existing tests, and I think I should drop the commit.  If we add a test, it should at least be in a different PR, due to its complexity.

I'll also start a question to libc maintainers about strtol(3), since we need to clarify the page regarding errno.

Thanks!

Reply:
To be addressed in a different PR